### PR TITLE
Provide logs from ActionRenderer APIs

### PR DESCRIPTION
**Pull request by longfin**
This PR extends (un)rendering APIs in `IActionRenderer`, to access `ActionEvaluation.Logs` data from the application side on action rendering.

it might look like that solves the same problem already covered in #2505 in a different way, and to some extent it does. However, from the application's perspective, if ephemeral execution logs generated during action execution can be obtained at rendering time, it has the advantage that it can be abstracted in a similar way to other action execution results without the need for separate access to `TxExecution`. (also there is a small performance gain because it doesn't require additional disk access 😅 )

Another approach that could be suggested would be to extend rendering API, to get whole`TxExecution` at render time as well as `logs`. but, since `TxExecution` doesn't seem to fit well with the hierarchy of the current rendering API, I think this approach needs a little different consideration.

Feel free to comment on the strategy as well as the implementation itself.

### CODE REVIEW

File: Libplanet/Blockchain/Renderers/Debug/RenderRecord.cs [hunk 1]
```diff
@@ -9,7 +9,7 @@ namespace Libplanet.Blockchain.Renderers.Debug
     /// </summary>
     /// <typeparam name="T">An <see cref="IAction"/> type.  It should match to
     /// <see cref="Libplanet.Blockchain.BlockChain{T}"/>'s type parameter.</typeparam>
-    public abstract class RenderRecord<T>
+public abstract class RenderRecord<T>
```
Comment (moreal) [inline Libplanet/Blockchain/Renderers/Debug/RenderRecord.cs:12]: This change seems not related to this pull request 🤔 This change should be reverted.

File: Libplanet/Blockchain/Renderers/NonblockActionRenderer.cs [hunk 2]
```diff
@@ -76,41 +77,49 @@ FullFallback fullFallback
...
+            IAccountStateDelta nextStates,
+            List<string> logs
         ) =>
-            Queue(() => ActionRenderer.RenderAction(action, context, nextStates));
+            Queue(() => ActionRenderer.RenderAction(action, context, nextStates, logs));
 
-        /// <inheritdoc
-        /// cref="IActionRenderer{T}.UnrenderAction(IAction, IActionContext, IAccountStateDelta)"/>
+#pragma warning disable MEN002
+        /// <inheritdoc cref="IActionRenderer{T}.UnrenderActionError(IAction, IActionContext, Exception, List{string})"/>
```
Comment (moreal) [inline Libplanet/Blockchain/Renderers/NonblockActionRenderer.cs:92]: Maybe this line should be like the below:
```suggestion
        /// <inheritdoc cref="IActionRenderer{T}.UnrenderAction(IAction, IActionContext, IAccountStateDelta, List{string})"/>
```

File: Libplanet/Blockchain/Renderers/LoggedActionRenderer.cs [hunk 3]
```diff
@@ -68,60 +69,68 @@ Block<T> newTip
...
                 action,
                 context,
-                () => ActionRenderer.RenderAction(action, context, nextStates)
+                () => ActionRenderer.RenderAction(action, context, nextStates, logs)
             );
 
-        /// <inheritdoc
-        /// cref="IActionRenderer{T}.UnrenderAction(IAction, IActionContext, IAccountStateDelta)"/>
+#pragma warning disable MEN002
+        /// <inheritdoc cref="IActionRenderer{T}.UnrenderActionError(IAction, IActionContext, Exception, List{string})"/>
```
Comment (moreal) [inline Libplanet/Blockchain/Renderers/LoggedActionRenderer.cs:89]:
```suggestion
        /// <inheritdoc cref="IActionRenderer{T}.UnrenderAction(IAction, IActionContext, IAccountStateDelta, List{string})"/>
```

File: Libplanet/Blockchain/Renderers/Debug/ValidatingActionRenderer.cs [hunk 4]
```diff
@@ -53,27 +53,31 @@ public override void RenderReorg(Block<T> oldTip, Block<T> newTip, Block<T> bran
             Validate();
         }
 
-        /// <inheritdoc
-        /// cref="IActionRenderer{T}.UnrenderAction(IAction, IActionContext, IAccountStateDelta)"/>
+#pragma warning disable MEN002
+        /// <inheritdoc cref="IActionRenderer{T}.UnrenderActionError(IAction, IActionContext, Exception, List{string})"/>
```
Comment (moreal) [inline Libplanet/Blockchain/Renderers/Debug/ValidatingActionRenderer.cs:57]:
```suggestion
        /// <inheritdoc cref="IActionRenderer{T}.UnrenderAction(IAction, IActionContext, IAccountStateDelta, List{string})"/>
```

File: Libplanet/Blockchain/Renderers/AtomicActionRenderer.cs [hunk 5]
```diff
@@ -88,20 +92,22 @@ IAccountStateDelta nextStates
...
             {
-                _eventBuffer.Add((action, context, nextStates));
+                _eventBuffer.Add((action, context, nextStates, logs));
             }
         }
 
-        /// <inheritdoc
-        /// cref="IActionRenderer{T}.UnrenderAction(IAction, IActionContext, IAccountStateDelta)"/>
+#pragma warning disable MEN002
+        /// <inheritdoc cref="IActionRenderer{T}.UnrenderActionError(IAction, IActionContext, Exception, List{string})"/>
```
Comment (moreal) [inline Libplanet/Blockchain/Renderers/AtomicActionRenderer.cs:104]:
```suggestion
        /// <inheritdoc cref="IActionRenderer{T}.UnrenderAction(IAction, IActionContext, IAccountStateDelta, List{string})"/>
```

File: Libplanet.Tests/Blockchain/BlockChainTest.cs [hunk 6]
```diff
@@ -392,7 +392,7 @@ public async Task RenderActionsAfterAppendComplete()
             var stateStore = new TrieStateStore(new MemoryKeyValueStore());
             IActionRenderer<DumbAction> renderer = new AnonymousActionRenderer<DumbAction>
             {
-                ActionRenderer = (_, __, nextStates) =>
+                ActionRenderer = (action_, ctx, nextStates, logs) =>
```
Comment (moreal) [inline Libplanet.Tests/Blockchain/BlockChainTest.cs:395]:
```suggestion
                ActionRenderer = (action, ctx, nextStates, logs) =>
```

Or only adding `logs` parameter:

```suggestion
                ActionRenderer = (_, __, nextStates, logs) =>
```

File: CHANGES.md [hunk 7]
```diff
@@ -16,6 +16,14 @@ To be released.
...
     `BlockCandidateTable<T>.GetCurrentRoundCandidate()` from `List<Block<T>>?`
     to `Branch<T>?`.  [[#2822]]
+ -  Added a `logs` parameter to the below methods in `IActionRenderer<T>`
+    interface and its implementations.  [[#2826]]
+    
+     - `IActionRenderer<T>.RenderAction()`.
+     - `IActionRenderer<T>.UnrenderAction()`.
+     - `IActionRenderer<T>.RenderActionError()`.
+     - `IActionRenderer<T>.UnrenderActionError()`.
+     - `IActionRenderer<T>.RenderAction()`.
```
Comment (moreal) [inline CHANGES.md:26]:
```suggestion
     - `IActionRenderer<T>.RenderAction()`.
     - `IActionRenderer<T>.UnrenderAction()`.
     - `IActionRenderer<T>.RenderActionError()`.
     - `IActionRenderer<T>.UnrenderActionError()`.
```

File: CHANGES.md [hunk 8]
```diff
@@ -16,6 +16,14 @@ To be released.
  -  (Libplanet.Net) Changed the return type of
     `BlockCandidateTable<T>.GetCurrentRoundCandidate()` from `List<Block<T>>?`
     to `Branch<T>?`.  [[#2822]]
+ -  Added a `logs` parameter to the below methods in `IActionRenderer<T>`
+    interface and its implementations.  [[#2826]]
```
Comment (moreal) [inline CHANGES.md:20]: We should add markdown link like the below:

```markdown
[#2826]: https://github.com/planetarium/libplanet/2826
```

Comment (longfin) [reply CHANGES.md:20]: Amended. aa67d5e7c

**Review by moreal — DISMISSED** [review]

**Review by dahlia — APPROVED** [review]
